### PR TITLE
Fixing prank: no longer override the sender on lower frames

### PR DESCRIPTION
**Pull request by msooseth**
## Description
As per #608 there is an issue with prank. It keeps overriding the sender on lower frames as well. This should not happen.

## Checklist

- [x] tested locally
- [ ] added automated tests
- [ ] updated the docs
- [x] updated the changelog

### CODE REVIEW

**Comment by msooseth** [comment]
@elopez could you maybe review?  Maybe it'd be useful to add more tests in `tests.hs`, perhaps you could have a go at that?

File: src/EVM.hs [hunk 1]
```diff
@@ -83,6 +83,8 @@ blankState = do
     , gas          = initialGas
     , returndata   = mempty
     , static       = False
+    , overrideCaller = Nothing
+    , resetCaller  = True
```
Comment (elopez) [inline src/EVM.hs:87]:
```suggestion
    , resetCaller  = False
```
to match with the override = Nothing

File: src/EVM.hs [hunk 2]
```diff
@@ -145,6 +147,8 @@ makeVm o = do
       , gas = o.gas
       , returndata = mempty
       , static = False
+      , overrideCaller = Nothing
+      , resetCaller = True
```
Comment (elopez) [inline src/EVM.hs:151]:
```suggestion
      , resetCaller = False
```
to match with the override = Nothing

File: src/EVM.hs [hunk 3]
```diff
@@ -866,17 +868,16 @@ exec1 = do
...
-                            let from' = fromMaybe self vm.config.overrideCaller
+                            let from' = fromMaybe self overrideC
                             zoom #state $ do
                               assign #callvalue xValue
                               assign #caller from'
                               assign #contract callee
-                            do
-                              resetCaller <- use (#config % #resetCaller)
-                              when (resetCaller) $ assign (#config % #overrideCaller) Nothing
+                              assign #overrideCaller Nothing
```
Comment (elopez) [inline src/EVM.hs:880]:
```suggestion
                              assign #overrideCaller Nothing
                              assign #resetCaller False                              
```

File: src/EVM.hs [hunk 4]
```diff
@@ -889,14 +890,13 @@ exec1 = do
...
+                    overrideC <- use $ #state % #overrideCaller
                     delegateCall this gas xTo self xValue xInOffset xInSize xOutOffset xOutSize xs $ \_ -> do
                       zoom #state $ do
                         assign #callvalue xValue
-                        assign #caller $ fromMaybe self vm.config.overrideCaller
-                      do
-                        resetCaller <- use (#config % #resetCaller)
-                        when (resetCaller) $ assign (#config % #overrideCaller) Nothing
+                        assign #caller $ fromMaybe self overrideC
+                        assign #overrideCaller Nothing
```
Comment (elopez) [inline src/EVM.hs:899]:
```suggestion
                        assign #overrideCaller Nothing
                        assign #resetCaller False
```

File: src/EVM.hs [hunk 5]
```diff
@@ -978,17 +978,16 @@ exec1 = do
...
                           zoom #state $ do
                             assign #callvalue (Lit 0)
-                            assign #caller $ fromMaybe self (vm.config.overrideCaller)
+                            assign #caller $ fromMaybe self overrideC
                             assign #contract callee
                             assign #static True
-                          do
-                            resetCaller <- use (#config % #resetCaller)
-                            when (resetCaller) $ assign (#config % #overrideCaller) Nothing
+                            assign #overrideCaller Nothing
```
Comment (elopez) [inline src/EVM.hs:990]:
```suggestion
                            assign #overrideCaller Nothing
                            assign #resetCaller False
```

File: src/EVM.hs [hunk 6]
```diff
@@ -1101,7 +1100,9 @@ callChecks this xGas xContext xTo xValue xInOffset xInSize xOutOffset xOutSize x
     accessMemoryRange xOutOffset xOutSize $ do
       availableGas <- use (#state % #gas)
       let recipientExists = accountExists xContext vm
-      let from = fromMaybe vm.state.contract vm.config.overrideCaller
+      let from = fromMaybe vm.state.contract vm.state.overrideCaller
+      resetCaller <- use $ #state % #resetCaller
+      when resetCaller $ assign (#state % #overrideCaller) Nothing
```
Comment (elopez) [inline src/EVM.hs:1105]: This should probably also set resetCaller to False at the same time

Comment (elopez) [reply src/EVM.hs:1105]: having this on callChecks feels a bit odd to me, should it maybe live on delegateCall instead?

File: src/EVM.hs [hunk 7]
```diff
@@ -1799,26 +1797,26 @@ cheatActions = Map.fromList
       \sig input -> case decodeStaticArgs 0 1 input of
         [addr]  -> case wordToAddr addr of
           Just a -> do
-            assign (#config % #overrideCaller) (Just a)
             doStop
```
Comment (elopez) [inline src/EVM.hs:1800]: Can you add a comment explaining why doStop is done first? I imagine this lets you edit the parent's `#state` which is the one who should hold the prank, but it's non-obvious.

File: src/EVM.hs [hunk 8]
```diff
@@ -1799,26 +1797,26 @@ cheatActions = Map.fromList
...
+            assign (#state % #resetCaller) False
           Nothing -> vmError (BadCheatCode "startPrank(address), could not decode address provided" sig)
         _ -> vmError (BadCheatCode "startPrank(address) parameter decoding failed" sig)
 
   , action "stopPrank()" $
       \_ _ -> do
-        assign (#config % #overrideCaller) Nothing
-        assign (#config % #resetCaller) True
         doStop
+        assign (#state % #overrideCaller) Nothing
```
Comment (elopez) [inline src/EVM.hs:1819]:
```suggestion
        assign (#state % #overrideCaller) Nothing
        assign (#state % #resetCaller) False
```

File: src/EVM.hs [hunk 9]
```diff
@@ -2014,6 +2011,7 @@ delegateCall this gasGiven xTo xContext xValue xInOffset xInSize xOutOffset xOut
                     assign #memorySize 0
                     assign #returndata mempty
                     assign #calldata calldata
+                    assign #overrideCaller Nothing
```
Comment (elopez) [inline src/EVM.hs:2014]:
```suggestion
                    assign #overrideCaller Nothing
                    assign #resetCaller False
```

**Review by elopez — COMMENTED** [review]
Overall it looks good to me! I left you a few suggestions to make resetCaller state more consistent

File: src/EVM.hs (hunk 7 above)
Comment (msooseth) [reply src/EVM.hs:1803]: This has been fixed now :)

File: src/EVM.hs (hunk 6 above)
Comment (msooseth) [reply src/EVM.hs:1108]: Hmmm... yeah, we could set `resetCaller` in `delegateCall`. However, the `let from = ..` needs to be here of course. Wanna have a go at putting it in `delegateCall`? Just create a PR on top of this PR :)

Comment (msooseth) [reply src/EVM.hs]: Thanks, good point! I fixed it :) Let me know what you think!

**Pull request merged**